sdk/go/common/workspace/plugins_test.go

```diff
@@ -1691,3 +1691,155 @@ func TestProjectPluginsWithSymlink(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, filepath.Join(tempdir, "symlink", "pulumi-resource-aws"), path)
 }
+
+func TestNewPluginSpec(t *testing.T) {
+	t.Parallel()
+
+	v1 := semver.MustParse("1.0.0")
+	v0deadbeef := semver.MustParse("0.0.0-xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef")
+	cases := []struct {
+		name               string
+		source             string
+		version            *semver.Version
+		kind               apitype.PluginKind
+		ExpectedPluginSpec PluginSpec
+		Error              error
+	}{
+		{
+			name:   "regular plugin",
+			source: "pulumi-example",
+			kind:   apitype.ResourcePlugin,
+			ExpectedPluginSpec: PluginSpec{
+				Name:              "pulumi-example",
+				Kind:              apitype.ResourcePlugin,
+				Version:           nil,
+				PluginDownloadURL: "",
+				PluginDir:         "",
+				Checksums:         nil,
+			},
+		},
+		{
+			name:   "plugin with version",
+			source: "pulumi-example@v1.0.0",
+			kind:   apitype.ResourcePlugin,
+			ExpectedPluginSpec: PluginSpec{
+				Name:              "pulumi-example",
+				Kind:              apitype.ResourcePlugin,
+				Version:           &v1,
+				PluginDownloadURL: "",
+				PluginDir:         "",
+				Checksums:         nil,
+			},
+		},
+		{
+			name:   "plugin with invalid semver",
+			source: "pulumi-example@v1.0.0.0",
+			kind:   apitype.ResourcePlugin,
+			Error:  errors.New("VERSION must be valid semver: Invalid character(s) found in patch number \"0.0\""),
+		},
+		{
+			name:   "git plugin",
+			source: "github.com/pulumi/pulumi-example",
+			kind:   apitype.ResourcePlugin,
+			ExpectedPluginSpec: PluginSpec{
+				Name:              "github.com_pulumi_pulumi-example",
+				Kind:              apitype.ResourcePlugin,
+				Version:           nil,
+				PluginDownloadURL: "github.com/pulumi/pulumi-example",
+				PluginDir:         "",
+				Checksums:         nil,
+			},
+		},
+		{
+			name:   "git plugin with version",
+			source: "github.com/pulumi/pulumi-example@v1.0.0",
+			kind:   apitype.ResourcePlugin,
+			ExpectedPluginSpec: PluginSpec{
+				Name:              "github.com_pulumi_pulumi-example",
```

Comment:
I'm not sure about these names. I feel like the plugin name should still be a simple name, like pulumi-example here and just plugin in the local dir tests below.

Reply:
See https://github.com/pulumi/pulumi/pull/17827#discussion_r1854068149